lib/internal/url.js

```diff
@@ -1365,27 +1366,54 @@ const backslashRegEx = /\\/g;
 const newlineRegEx = /\n/g;
 const carriageReturnRegEx = /\r/g;
 const tabRegEx = /\t/g;
+
+function encodePathChars(filepath) {
+  if (filepath.includes('%'))
+    filepath = filepath.replace(percentRegEx, '%25');
+  // In posix, "/" is a valid character in paths
+  if (!isWindows && filepath.includes('\\'))
+    filepath = filepath.replace(backslashRegEx, '%5C');
+  if (filepath.includes('\n'))
+    filepath = filepath.replace(newlineRegEx, '%0A');
+  if (filepath.includes('\r'))
+    filepath = filepath.replace(carriageReturnRegEx, '%0D');
+  if (filepath.includes('\t'))
+    filepath = filepath.replace(tabRegEx, '%09');
+  return filepath;
+}
+
 function pathToFileURL(filepath) {
-  let resolved = path.resolve(filepath);
-  // path.resolve strips trailing slashes so we must add them back
-  const filePathLast = filepath.charCodeAt(filepath.length - 1);
-  if ((filePathLast === CHAR_FORWARD_SLASH ||
-       (isWindows && filePathLast === CHAR_BACKWARD_SLASH)) &&
-      resolved[resolved.length - 1] !== path.sep)
-    resolved += '/';
   const outURL = new URL('file://');
-  if (resolved.includes('%'))
-    resolved = resolved.replace(percentRegEx, '%25');
-  // In posix, "/" is a valid character in paths
-  if (!isWindows && resolved.includes('\\'))
-    resolved = resolved.replace(backslashRegEx, '%5C');
-  if (resolved.includes('\n'))
-    resolved = resolved.replace(newlineRegEx, '%0A');
-  if (resolved.includes('\r'))
-    resolved = resolved.replace(carriageReturnRegEx, '%0D');
-  if (resolved.includes('\t'))
-    resolved = resolved.replace(tabRegEx, '%09');
-  outURL.pathname = resolved;
+  if (isWindows && filepath.startsWith('\\\\')) {
+    // UNC path format: \\server\share\resource
+    const paths = filepath.split('\\');
+    if (paths.length <= 3) {
```

Comment:
Shouldn't this be `paths.length <= 4`? Or can resource be omitted?

```
> '\\\\server\\share\\resource'.split('\\')
[ '', '', 'server', 'share', 'resource' ]
```

Reply:
If resource is omitted, it's the root directory of the share.